mailbox/store/src/main/java/org/apache/james/mailbox/store/MessageStorer.java

```diff
@@ -72,22 +73,25 @@
         private final MessageFactory messageFactory;
         private final AttachmentMapperFactory attachmentMapperFactory;
         private final MessageParser messageParser;
+        private final ThreadIdGuessingAlgorithm threadIdGuessingAlgorithm;
 
         public WithAttachment(MailboxSessionMapperFactory mapperFactory, MessageId.Factory messageIdFactory,
                               MessageFactory messageFactory, AttachmentMapperFactory attachmentMapperFactory,
-                              MessageParser messageParser) {
+                              MessageParser messageParser, ThreadIdGuessingAlgorithm threadIdGuessingAlgorithm) {
             this.mapperFactory = mapperFactory;
             this.messageIdFactory = messageIdFactory;
             this.messageFactory = messageFactory;
             this.attachmentMapperFactory = attachmentMapperFactory;
             this.messageParser = messageParser;
+            this.threadIdGuessingAlgorithm = threadIdGuessingAlgorithm;
         }
 
         @Override
         public Mono<Pair<MessageMetaData, Optional<List<MessageAttachmentMetadata>>>> appendMessageToStore(Mailbox mailbox, Date internalDate, int size, int bodyStartOctet, Content content, Flags flags, PropertyBuilder propertyBuilder, Optional<Message> maybeMessage, MailboxSession session) throws MailboxException {
             MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
             MessageId messageId = messageIdFactory.generate();
-            ThreadId threadId = ThreadId.fromBaseMessageId(messageId);
+            // TODO get mime message header fields
```

Comment:
I think you need a hint there...

MaximalBodyDescriptor do held a list of fields CF this debugging session:

![Screenshot from 2021-06-29 16-17-11](https://user-images.githubusercontent.com/6928740/123771846-95fe7300-d8f5-11eb-841a-b0fb0aa600ae.png)

We can contribute a `MaximalBodyDescriptor::getFiled(String name)` method in mime4J ?

Then we would likely be able to do threadId guessing easily in `StoreMessageManager::appendMessage` (the one with the big try catch...) Then we can pass the threadId as a parameter to other methods just like properties...

Reply:
Likely even better (and less hacky) in arseProperties there is a header isntance we do not use...

```
private PropertyBuilder parseProperties(BodyOffsetInputStream bIn) throws IOException, MimeException {
        // Disable line length... This should be handled by the smtp server
        // component and not the parser itself
        // https://issues.apache.org/jira/browse/IMAP-122
        final MimeTokenStream parser = getParser(bIn);
        final HeaderImpl fields = readHeader(parser);
        
        // All we need for threadId guessing <3
        fields.getFields("Message-Id")
        
        final MaximalBodyDescriptor descriptor = (MaximalBodyDescriptor) parser.getBodyDescriptor();
        final MediaType mediaType = getMediaType(descriptor);
        final PropertyBuilder propertyBuilder = getPropertyBuilder(descriptor, mediaType.mediaType, mediaType.subType);
        setTextualLinesCount(parser, mediaType.mediaType, propertyBuilder);
        return propertyBuilder;
    }
 ```

Reply:
How about using this header instance to get headers fields needed for ThreadIdGuessingAlgorithm, and still place ThreadIdGuessingAlgorithm at MessageStorer::appendMessageToStore as I am doing now? Because I think threadId should go with messageId. Otherwise we should messageId generate out of MessageStorer::appendMessageToStore.